### PR TITLE
cluster: add tt cluster command

### LINKED ISSUE #506: cluster: add the ability to manipulate the configuration
In the scope of the task needs to add the ability to manipulate configuration (show / update) of single instance and cluster (config from etcd) according to https://www.notion.so/tarantool/tt-3-0-fbd40e724dce4a04919d1a35c506a65b

### PR DESCRIPTION
**Pull request by oleg-jukovec**
* `tt cluster show` to show a configuration.
* `tt cluster publish` to publish a configuration.

@TarantoolBot document
Title: new module - `cluster`

The patch introduces a new module `cluster`. The module has two commands:

* `show` - to show a configuration for a cluster or an instance.
* `publish` - to publish a cluster or an instance configuration.

The target could be a cluster or an instance configuration. It could be a fully-merged (from all sources) configuration for an application or just a remote configuration part in case of etcd.

Usage:

* `tt cluster show (<APP_NAME> | <APP_NAME:INSTANCE_NAME> | <URI>) [--validate]`
  * The optional `--validate` flag enables validation for a showed configuration.
* `tt cluster publish (<APP_NAME> | <APP_NAME:INSTANCE_NAME> | <URI>) file [--force]`
  * The optional `--force` flag skips validation of uploaded configuration.

The `APP_NAME` is a name of an application.

The `INSTANCE_NAME` is a name of an instance in the application.

The `URI` specifies a etcd connection settings in the following format:

`http(s)://[username:password@]host:port[/prefix][?arguments]`
    
* `prefix` - a base path to Tarantool configuration in etcd.

Possible arguments:

* `name` - a name of an instance in the cluster configuration.
* `timeout` - a request timeout in seconds (default 3.0).
* `ssl_key_file` - a path to a private SSL key file.
* `ssl_cert_file` - a path to an SSL certificate file.
* `ssl_ca_file` - a path to a trusted certificate authorities (CA) file.
* `ssl_ca_path` - a path to a trusted certificate authorities (CA) directory.
* `verify_host` - set off (default true) verification of the certificate’s name against the host.
* `verify_peer` - set off (default true) verification of the peer’s SSL certificate.

Closes #506

### CODE REVIEW

**Comment by oleg-jukovec** [comment]
An example of a pull request from the CI run: https://github.com/tarantool/tt/pull/591

File: cli/cluster/etcd.go [hunk 1]
```diff
@@ -196,6 +195,99 @@ func (collector EtcdCollector) Collect() (*Config, error) {
...
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		resp, err := publisher.getter.Get(ctx, prefix, clientv3.WithPrefix())
+		if err != nil {
+			return fmt.Errorf("failed to fetch data from etcd: %w", err)
+		}
+
```
Comment (psergee) [inline cli/cluster/etcd.go:249]: Please add a comment with the description of the operations below.

**Review by psergee — APPROVED** [review]

File: cli/cmd/cluster.go [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+				running.ExtractActiveAppNames,
+				running.ExtractActiveInstanceNames)
+		},
+	}
+	show.Flags().BoolVar(&showFlags.Validate, "validate", showFlags.Validate,
+		"validate the configuration")
+	clusterCmd.AddCommand(show)
+
+	public := &cobra.Command{
+		Use:   "public (<APP_NAME> | <APP_NAME:INSTANCE_NAME> | <URI>) file",
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go:100]: It is confusing me too. We could discuss at next sync.

File: cli/cluster/config.go [hunk 3]
```diff
@@ -87,7 +87,12 @@ func (config *Config) Set(path []string, value any) error {
 	if m, err := config.createMaps(path[0:last]); err != nil {
 		return err
 	} else {
-		m[path[last]] = value
+		key := path[last]
+		if cfg, ok := value.(*Config); value != nil && ok {
+			m[key] = cfg.paths
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cluster/config.go]: Added `deepCopy` of maps to the code.

File: cli/cluster/config_test.go [hunk 4]
```diff
@@ -42,6 +42,27 @@ func TestConfig_Set(t *testing.T) {
...
+	require.NoError(t, err)
+
+	err = base.Set([]string{"foo", "bar", "zoo"}, add)
+	require.NoError(t, err)
+
+	valueAdd, err := add.Get([]string{"foo", "bar"})
+	require.NoError(t, err)
+	assert.Equal(t, 2, valueAdd)
+
+	valueAdded, err := base.Get([]string{"foo", "bar", "zoo", "foo", "bar"})
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cluster/config_test.go]: The test updated.

File: cli/cluster/paths_generate.lua [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+-- Sort in the alphabet order to avoid mixing on a next generation.
+table.sort(configs, function(left, right)
+    local pathleft = table.concat(left.path, "_")
+    local pathright = table.concat(right.path, "_")
+    return pathleft < pathright
+end)
+
+-- The file generation.
+local file = fio.open("paths.go", {'O_WRONLY', 'O_CREAT', 'O_TRUNC'})
+file:write([[// Code generated by env_generate.lua; DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cluster/paths_generate.lua]: Thank you, fixed.

File: cli/cluster/validators.go [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+	return err.path
+}
+
+// Error returns a string representation of the validation error.
+func (err ValidateError) Error() string {
+	if len(err.errs) < 1 {
+		return fmt.Sprintf("failed to validate %q", err.path)
+	}
+
+	return fmt.Sprintf("failed to validate %q: %s",
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cluster/validators.go]: Changed to: `invalid path %q: %s`

File: magefile.go [hunk 7]
```diff
@@ -302,7 +302,8 @@ func IntegrationNoTarantool() error {
 func Codespell() error {
 	fmt.Println("Running codespell tests...")
 
-	return sh.RunV("codespell", packagePath, "test", "README.md", "doc", "CHANGELOG.md")
+	return sh.RunV("codespell", packagePath, "test", "README.md", "doc", "CHANGELOG.md",
+		"-L", "ro", "--skip=./cli/cluster/paths.go")
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply magefile.go]: Updated.

File: cli/cmd/cluster.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: `public`->`publish`

File: cli/cmd/cluster.go [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+	prefix, timeout := etcdOpts.Prefix, etcdOpts.Timeout
+	config, err := cluster.NewEtcdCollector(etcdcli, prefix, timeout).Collect()
+	if err != nil {
+		return fmt.Errorf("failed to collect a configuration from etcd: %w", err)
+	}
+
+	return printRawClusterConfig(config, uri.Query().Get("name"), validate)
+}
+
+// showFile shows a full cluster configuration for a configuration path.
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: The code moved to `cli/cluster/cmd/show.go`, updated.

File: cli/cmd/cluster.go [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+	config, err := cluster.GetClusterConfig(path)
+	if err != nil {
+		return fmt.Errorf("failed to get a cluster configuration: %w", err)
+	}
+
+	return printClusterConfig(config, name, validate)
+}
+
+// internalClusterPublicModule is an entrypoint for `cluster public` command.
+func internalClusterPublicModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: The commands code moved to `cli/cluster/cmd/`.
I don't sure that this is a good idea to mix a cluster configuration functionality with a user (potentially interactive) commands. So I've created a separate subpackage just for the commands code.

File: cli/cmd/cluster.go [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+	}
+
+	if validate {
+		err = validateInstanceConfig(iconfig, instance)
+	}
+	printConfig(iconfig)
+	return err
+}
+
+// vlidateRawConfig validates a raw cluster or an instance configuration. The
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: Thank you, updated.

File: cli/cmd/cluster.go [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+	cconfig, err := cluster.MakeClusterConfig(config)
+	if err != nil {
+		return err
+	}
+
+	return validateClusterConfig(cconfig, false)
+}
+
+// validateClusterConfig validates a cluster configuration.
+func validateClusterConfig(cconfig cluster.ClusterConfig, full bool) error {
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: Added, see `cli/cluster/cmd/common_test.go`.

File: cli/cmd/cluster.go [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+	colonIds := strings.Index(appStr, string(running.InstanceDelimiter))
+	if colonIds != -1 {
+		name = runningCtx.Instances[0].InstName
+	}
+
+	return path, name, nil
+}
+
+// parseEtcdOpts parses etcd options from a URL.
+func parseEtcdOpts(uri *url.URL) (cluster.EtcdOpts, error) {
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: Added, see `cli/cluster/etcd_test.go:TestMakeEtcdOptsFromUrl`.

File: cli/cluster/etcd.go (hunk 1 above)
Comment (oleg-jukovec) [reply cli/cluster/etcd.go:268]: I've added more comments to the code.

File: test/integration/cluster/test_cluster.py [hunk 13]
```diff
@@ -0,0 +1,790 @@
...
+        instances:
+          master:
+            database:
+              mode: rs
+              txn_timeout: asd
+            iproto:
+              listen: 127.0.0.1:3301
+   ⨯ an invalid instance "master" configuration:"""
+    expected += r""" failed to validate ["database" "mode" "rs"]: should be one of [ro rw]
+failed to validate ["database" "txn_timeout" "asd"]: failed to parse number
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply test/integration/cluster/test_cluster.py]: The format of the error changed:

`invalid path ["path" "to" "value"]: an error message probably with "value value"`

as example:

`invalid path ["audit_log" "nonblock"]: unexpected value "123" of type string, expected boolean`

File: .github/workflows/update.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
+name: update
+
+on:
+  schedule:
+    - cron:  '0 0 * * *'
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply .github/workflows/update.yml:7]: Added.

File: cli/cmd/cluster.go [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+		return "", "", fmt.Errorf("unable to resolve the application name %q: %w",
+			appStr, errNoConfig)
+	}
+
+	err := running.FillCtx(cliOpts, cmdCtx, &runningCtx, []string{appStr})
+	if err != nil {
+		return "", "", err
+	}
+
+	path := filepath.Join(runningCtx.Instances[0].AppDir, configFileName)
```
Comment (psergee) [inline cli/cmd/cluster.go:198]: Let's use `ClusterConfigPath`.
`path` variable name purpose is unclear. Must be "clusterConfigPath" or something like this.

Comment (oleg-jukovec) [reply cli/cmd/cluster.go]: Updated.

**Pull request merged**